sdk/core/core/inc/az_span.h

```diff
@@ -135,161 +158,306 @@ AZ_NODISCARD az_span az_span_from_str(char* str);
  */
-AZ_NODISCARD az_result az_span_to_uint64(az_span self, uint64_t* out);
+AZ_NODISCARD az_result
+az_span_to_str(char* destination, int32_t destination_max_size, az_span source);
+
+/******************************  SPAN PARSING */
 
 /**
- * @brief convert all the span content to uint32
+ * @brief az_span_to_uint64 parses an az_span containing ASCII digits into a uint64 number
  *
- * @param self source span
- * @param out result span
- * @return AZ_NODISCARD az_span_to_uint32
+ * @param[in] span The az_span containing the ASCII digits to be parsed.
+ * @param[in] out_number The pointer to the variable that is to receive the number
+ * @return An #az_result value indicating the result of the operation.
+ *          #AZ_OK if successful
+ *          #AZ_ERROR_PARSER_UNEXPECTED_CHAR if a non-ASCII digit is found within the span.
  */
-AZ_NODISCARD az_result az_span_to_uint32(az_span self, uint32_t* out);
+
+AZ_NODISCARD az_result az_span_to_uint64(az_span span, uint64_t* out_number);
 
 /**
- * @brief move the content of span @b src to @b dst
+ * @brief az_span_to_uint32 parses an az_span containing ASCII digits into a uint32 number
  *
- * @param dst buffer where to copy span
- * @param src span to be copied
- * @param out span resulting after copy
- * @return AZ_NODISCARD az_span_move
+ * @param span The az_span containing the ASCII digits to be parsed.
+ * @param out_number The pointer to the variable that is to receive the number
+ * @return An #az_result value indicating the result of the operation.
+ *          #AZ_OK if successful
+ *          #AZ_ERROR_PARSER_UNEXPECTED_CHAR if a non-ASCII digit
+ * is found within the span.
  */
-AZ_NODISCARD az_result az_span_copy(az_span dst, az_span src, az_span* out);
+AZ_NODISCARD az_result az_span_to_uint32(az_span span, uint32_t* out_number);
 
-AZ_NODISCARD az_result az_span_copy_url_encode(az_span dst, az_span src, az_span* out);
+/******************************  SPAN APPENDING */
 
 /**
- * @brief append az_span if there is enough capacity for it
+ * @brief az_span_append_uint8 appends the uint8 a to the destination starting at the
+ * destination span's length.
  *
- * @param self src span where to append
- * @param span content to be appended
- * @return AZ_NODISCARD az_span_append_
+ * @param[in] destination The az_span where the byte should be appended to.
+ * @param[in] byte The uint8 to append to the destination span
+ * @param[out] out_span A pointer to an az_span that receives the span referring to the
+ * destination span with its length increased by 1
+ * @return An #az_result value indicating the result of the operation.
+ *          #AZ_OK if successful
+ *          #AZ_ERROR_INSUFFICIENT_SPAN_CAPACITY if the destination is not big enough to
+ * contain the appended byte
  */
-AZ_NODISCARD az_result az_span_append(az_span self, az_span span, az_span* out);
+AZ_NODISCARD az_result az_span_append_uint8(az_span destination, uint8_t byte, az_span* out_span);
```

Comment:
This API shape is confusing to me and the `out_span` parameter description doesn't help :(
What's `out_span` supposed to contain and how is it used?

Reply:
This is a very common pattern for us. The destination span is passed by value with its pointer, length, & capacity. The returned span has the same pointer & capacity but the length is different based on what got appended. This allows all span instance to be "immutable" which reduces the chance of memory corruption/race conditions occurring. The end result is a much more predictable code base. And, if we ever do have a bug, it should be much simpler to locate to fix.

Reply:
Alright. Maybe we can reword the `out_span` description then. I'll see if I can propose something that I would find helpful as a reader.